### PR TITLE
In-built modules can be indexed with the dot operator

### PR DESCRIPTION
**Pull request by itssamuelrowe**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -9,3 +9,7 @@ lua/src/luac
 cbenchmarks/*
 !cbenchmarks/*.c
 !cbenchmarks/makefile
+
+test.pln
+test.so
+test.lua
```
Comment (hugomg) [inline .gitignore:15]: We already have a gitignore rule to ignore ".so" files.

File: pallene/ast.lua [hunk 2]
```diff
@@ -83,7 +83,15 @@ declare_type("Field", {
 -- note: the following functions are why we need `if type(conss) == "table"` in parser.lua
 --
 
--- Return the variables names declared by a given toplevel node
+-- Returns a sequence containing the variable names declared by the specified
+-- toplevel node.
```
Comment (hugomg) [inline pallene/ast.lua:87]: It looks like your text editor is configured with lines of length 80?

File: pallene/ast.lua [hunk 3]
```diff
@@ -83,7 +83,15 @@ declare_type("Field", {
...
+--
+-- The following nodes will result in a sequence with one element.
+--     ast.Toplevel.Func, ast.Toplevel.Typealias, ast.Toplevel.Typealias,
+--     ast.Toplevel.Record, ast.Toplevel.Import, and ast.Toplevel.Builtin 
+-- The following nodes will result in a sequence with one or more elements.
+--     ast.Toplevel.Var
+--
```
Comment (hugomg) [inline pallene/ast.lua:94]: I don't know if we need this part of the comment. In the places that use ast.toplevel_names the user should not care whether each case is returning one or more names.

File: pallene/checker.lua [hunk 4]
```diff
@@ -53,25 +54,34 @@ function checker.check(prog_ast)
...
+-- and report a Pallene compilation error.
+local function checker_error(error_location, format, ...)
+    local error_message = location.format_error(error_location, format, ...)
+    coroutine.yield(error_message)
 end
 
-local function scope_error(loc, fmt, ...)
-    return checker_error(loc, ("scope error: " .. fmt), ...)
+local function scope_error(error_location, format, ...)
+    return checker_error(error_location, ("scope error: " .. format), ...)
```
Comment (hugomg) [inline pallene/checker.lua:72]: We are already using `loc` and `fmt` in other places in the code. I think it would be more consistent to keep the old names.

File: pallene/checker.lua [hunk 5]
```diff
@@ -198,31 +212,63 @@ local letrec_groups = {
...
+--    a flat `ast.Var.Name`.
+--
+--
+-- In the AST, unions tagged with `the ast.Var` refer to things that can appear
+-- on the left-hand side of an assignment. That is, variable names such as
+-- `foo` (`ast.Var.Name`), bracketed expressions such as `foo[i]` (`ast.Var.Bracket`)
+-- and qualified names such as `foo.bar` (`ast.Var.Dot`).
+-- In a union tagged as `ast.Stat.Assign`, the `vars` field refers to the `ast.Var`
+-- nodes in the left-hand-side of the assignment statement. Since Pallene has multiple
+-- assignments there may be one or more of those nodes.
```
Comment (hugomg) [inline pallene/checker.lua:236]: The right place to put comments about the meaning of the `ast.*` data structures would be in the ast.lua file. That said, I am not sure that this particular paragraph is saying something that is not already being said by the grammar itself.

File: pallene/checker.lua [hunk 6]
```diff
@@ -198,31 +212,63 @@ local letrec_groups = {
...
 function Checker:check_program(prog_ast)
 
     do
-        -- Forbid toplevel duplicates
+        -- Forbid top-level duplicates
+        --
+        -- Retrieve the names for all the top-level components such as functions,
+        -- imports, builtins and variables. Construct an auxillary table with name and location
+        -- as key and value, respectively. If the current name already exists in the auxillary
+        -- table, report an error.
```
Comment (hugomg) [inline pallene/checker.lua:246]: You could change this comment to say the  "why" instead of the "how".  Instead of talking about the auxiliary table, which is an implementation detail, it should say that we block duplicates to avoid ambiguities when exporting functions. We could also add a link to issue #184.

File: pallene/checker.lua [hunk 7]
```diff
@@ -587,8 +637,11 @@ function FunChecker:check_var(var)
             var._type = self.p.module.functions[cname.id].typ
         elseif cname._tag == "checker.Name.Builtin" then
             var._type = builtins[cname.name].typ
-               else
-            error("impossible")
+        elseif cname._tag == "checker.Name.Module" then
+            -- The type for a module is a table.
+            var._type = builtin_modules[cname.name]
+        else
+            error("Unexpected condition: cname._tag == " .. cname._tag)
```
Comment (hugomg) [inline pallene/checker.lua:644]: Let's leave this as `error("impossible")` for now. I think it will be cleaner if we make this kind of change in a separate PR and if we change all the `error("imposible")`s at once.

File: pallene/ast.lua (hunk 2 above)
Comment (hugomg) [reply pallene/ast.lua:87]: (Adding editorconfig support to your editor should take care of this)

File: .gitignore (hunk 1 above)
Comment (itssamuelrowe) [reply .gitignore:15]: I added these lines for development. I'll remove them before we merge.

File: pallene/ast.lua (hunk 3 above)
Comment (itssamuelrowe) [reply pallene/ast.lua:94]: Fair enough. I'll remove it.

File: pallene/checker.lua (hunk 4 above)
Comment (itssamuelrowe) [reply pallene/checker.lua:72]: Alright. I'll revert the changes I made.

File: pallene/checker.lua (hunk 5 above)
Comment (itssamuelrowe) [reply pallene/checker.lua:236]: > The right place to put comments about the meaning of the ast.* data structures would be in the ast.lua file.
I'll move it.

> That said, I am not sure that this particular paragraph is saying something that is not already being said by the grammar itself.
WIth that logic, our code wouldn't need any comments at all. :P
I'll remove it if you want though.

File: pallene/checker.lua (hunk 6 above)
Comment (itssamuelrowe) [reply pallene/checker.lua:246]: Alright.

File: pallene/checker.lua (hunk 7 above)
Comment (itssamuelrowe) [reply pallene/checker.lua:644]: Okay. I'll revert it.

File: .gitignore (hunk 1 above)
Comment (hugomg) [reply .gitignore:15]: You could consider using `git add -u` instead of plain `git add`. That way it only adds changes to files that are already in the repository and ignores the extra test files that you create. No need to change the gitignore file.

File: pallene/checker.lua [hunk 8]
```diff
@@ -11,7 +11,7 @@ local symtab = require "pallene.symtab"
 local types = require "pallene.types"
 local typedecl = require "pallene.typedecl"
 local util = require "pallene.util"
-local builtin_modules = require "pallene.builtin_modules"
+local inspect = require "inspect"
```
Comment (hugomg) [inline pallene/checker.lua:14]: Tip: you can require inspect in the same line where you use it. This way you don't forget about the `require "inspect"` at the top of the file after you delete the debugging print statement.

     print(require("inspect")(something))

File: pallene/builtins.lua [hunk 9]
```diff
@@ -4,24 +4,45 @@
...
-        typ = typ,
+builtins.functions = {
+    io_write = {
+        name = "io.write",
+        typ = T.Function({ T.String() }, {})
+    },
+    tofloat = {
+        name = "tofloat",
+        typ = T.Function({ T.Integer() }, { T.Float() }),
+    },
```
Comment (hugomg) [inline pallene/builtins.lua:19]: Do we need this `name` field? I suspect that we could just use key=name and value=type, as in the old builtins table:

    ["tofloat"] = T.Function({ T.Integer() }, { T.Float() })

File: pallene/checker.lua [hunk 10]
```diff
@@ -621,6 +621,8 @@ end
 function FunChecker:check_var(var)
     local tag = var._tag
     if     tag == "ast.Var.Name" then
+        -- print(inspect(var))
+        -- print('------------')
```
Comment (hugomg) [inline pallene/checker.lua:625]: Please delete these comments.

File: pallene/checker.lua [hunk 11]
```diff
@@ -636,29 +638,47 @@ function FunChecker:check_var(var)
         elseif cname._tag == "checker.Name.Function" then
             var._type = self.p.module.functions[cname.id].typ
         elseif cname._tag == "checker.Name.Builtin" then
-            var._type = builtins[cname.name].typ
+            var._type = builtins.functions[cname.name].typ
         elseif cname._tag == "checker.Name.Module" then
             -- The type for a module is a table.
-            var._type = builtin_modules[cname.name]
+           var._type = builtins.modules[cname.name]
```
Comment (hugomg) [inline pallene/checker.lua:644]: This case should be a type error, just like `checker.Name.Type`. This way, we will disallow things such as

    local x = io

By the way, don't forget to add a test case for that.

File: pallene/checker.lua [hunk 12]
```diff
@@ -636,29 +638,47 @@ function FunChecker:check_var(var)
...
         else
-            error("Unexpected condition: cname._tag == " .. cname._tag)
+            error("impossible")
         end
 
+        -- print(inspect(var))
```
Comment (hugomg) [inline pallene/checker.lua:649]: More print comments. Please remove all of them. (Consider using `git add -p`, which makes it possible to leave them in the text file but not include them in the commit)

File: pallene/checker.lua [hunk 13]
```diff
@@ -636,29 +638,47 @@ function FunChecker:check_var(var)
...
+            local function_name = var.name
+            local internal_name = module_name .. "." .. function_name
+            local cname = self.p.symbol_table:find_symbol(internal_name)
+            return {
+                _name = cname,
+                _tag = "ast.Var.Name",
+                loc = var.exp.loc,
+                name = var.exp.var.name .. '.' .. var.name,
+                _type = builtins.functions[internal_name].typ
+            }
```
Comment (hugomg) [inline pallene/checker.lua:665]: We should use the `ast.Var.Name` constructor instead of building this table by hand:

     local flat_var = ast.Var.Name(var.exp.loc, internal_name)
     flat_var._name = ...
     flat_var._type = ...

File: pallene/checker.lua [hunk 14]
```diff
@@ -636,29 +638,47 @@ function FunChecker:check_var(var)
...
-                types.tostring(ind_type))
-        end
-        local field_type = types.indices(ind_type)[var.name]
-        if not field_type then
-            type_error(var.loc,
-                "field '%s' not found in type '%s'",
-                var.name, types.tostring(ind_type))
+        -- print(inspect(var))
+        -- print('---------------------')
+        if builtins.modules[var.exp.var.name] ~= nil then
```
Comment (hugomg) [inline pallene/checker.lua:654]: Before you access `exp.var` you need to check that the `exp` in question is an `ast.Exp.Var`. This line of code will blow up if the expression on the left of the dot is not a variable name. For example `xs[i].name`

By the way, if the existing test suite is not catching this bug then please add a test case that ensures that it does :)

File: pallene/types.lua [hunk 15]
```diff
@@ -206,7 +206,9 @@ function types.consistent(t1, t2)
     return equivalent(t1, t2, true)
 end
 
+local inspect = require 'inspect'
 function types.tostring(t)
+    print(inspect(t))
```
Comment (hugomg) [inline pallene/types.lua:211]: This inspect is printing a lot of junk when we run `busted`.

File: pallene/checker.lua (hunk 14 above)
Comment (hugomg) [reply pallene/checker.lua:654]: Another thing: you can leave out the `~= nil`. In the rest of the code base we just use `if set[key] then` instead of `if set[key] ~= nil then`

File: pallene/ast.lua (hunk 2 above)
Comment (itssamuelrowe) [reply pallene/ast.lua:87]: I manually formatted it to 80 columns. I'll add the editorconfig support ASAP.

File: pallene/checker.lua (hunk 8 above)
Comment (itssamuelrowe) [reply pallene/checker.lua:14]: Alright. Will make the changes.

File: pallene/checker.lua (hunk 10 above)
Comment (itssamuelrowe) [reply pallene/checker.lua:625]: Sure.

File: pallene/checker.lua (hunk 11 above)
Comment (itssamuelrowe) [reply pallene/checker.lua:644]: Alright. Will start working on it tomorrow.

File: pallene/checker.lua (hunk 14 above)
Comment (itssamuelrowe) [reply pallene/checker.lua:654]: Done.

File: pallene/types.lua (hunk 15 above)
Comment (itssamuelrowe) [reply pallene/types.lua:211]: I must have pushed it without commenting. Sorry. Will remove it.

File: pallene/checker.lua [hunk 16]
```diff
@@ -586,27 +612,50 @@ function FunChecker:check_var(var)
...
-            type_error(var.loc,
-                "trying to access a member of value of type '%s'",
-                types.tostring(ind_type))
-        end
-        local field_type = types.indices(ind_type)[var.name]
-        if not field_type then
-            type_error(var.loc,
-                "field '%s' not found in type '%s'",
-                var.name, types.tostring(ind_type))
+        if var.exp._tag == "ast.Exp.Var" and builtins.modules[var.exp.var.name] then
```
Comment (hugomg) [inline pallene/checker.lua:627]: This also needs to check if `var.exp.var` is a `ast.Var.Name`. It turns out that the example I had previously provided where `var.exp.var` is a `var.Bracket` works by chance because `builtins.modules[nil]` is `nil`. However, this code will do weird things if `var.exp.var` is an `ast.Var.Dot` because that case also has a `.name` field.

File: pallene/checker.lua [hunk 17]
```diff
@@ -529,6 +554,7 @@ function FunChecker:check_stat(stat)
                         "attempting to assign to builtin function %s",
                         stat.vars[i].name)
                 end
+                -- TODO: Add checker.Name.Module to prevent module assignment
```
Comment (hugomg) [inline pallene/checker.lua:557]: Do we still need this TODO?

File: pallene/checker.lua [hunk 18]
```diff
@@ -407,10 +429,13 @@ function FunChecker:expand_function_returns(lhs, rhs)
...
 end
 
-
 function FunChecker:check_function(lambda, func_typ)
     assert(lambda._tag == "ast.Exp.Lambda")
 
+    -- Check for duplicate parameter names.
+    --
+    -- Add each parameter name to the `names` set. If the current name already
+    -- exists in the set, report an error.
```
Comment (hugomg) [inline pallene/checker.lua:438]: We don't need to add a comment here because PR #222 got rid of this check for duplicate parameter names. I'm worried that git will gladly merge this comment into the master branch even though the block of code that it is commenting won't exist anymore.

File: .gitignore [hunk 19]
```diff
@@ -8,4 +8,4 @@ lua/src/luac
 
 cbenchmarks/*
 !cbenchmarks/*.c
-!cbenchmarks/makefile
+!cbenchmarks/makefile
```
Comment (hugomg) [inline .gitignore:11]: We can make the diff a bit cleaner if we add back the newline.
